File: pkg/homekit_property.py
"""HomeKit adapter for Mozilla IoT Gateway."""

from gateway_addon import Property

from .util import hsv_to_rgb, rgb_to_hsv


class HomeKitProperty(Property):
    """HomeKit property type."""

    def __init__(self, device, aid, iid, name, description, value):
        """
        Initialize the object.

        device -- the Device this property belongs to
        aid -- the property's accessory id
        iid -- the property's instance id
        name -- name of the property
        description -- description of the property, as a dictionary
        value -- current value of this property
        """
        Property.__init__(self, device, name, description)
        self.aid = aid
        self.iid = iid
        self.set_cached_value(value)

    def set_value(self, value):
        """
        Set the current value of the property.

        value -- the value to set
        """
        ret = self.device.client.set_characteristics({
            '{}.{}'.format(self.aid, self.iid): value,
        })

        if ret:
            self.set_cached_value(value)
            self.device.notify_property_changed(self)

    def update(self, characteristics):
        """
        Update the current value, if necessary.

        characteristics -- current characteristic list for the device
        """
        c = list(
            filter(lambda x: x['aid'] == self.aid and x['iid'] == self.iid,
                   characteristics))
        if not c:
            return

        value = c[0]['value']
        if value != self.value:
            self.set_cached_value(value)
            self.device.notify_property_changed(self)


class HomeKitPlugProperty(HomeKitProperty):
    """Property type for HomeKit smart plugs."""

    def set_value(self, value):
        """
        Set the current value of the property.

        value -- the value to set
        """
        if self.name == 'on':
            HomeKitProperty.set_value(self, value)


class HomeKitBulbProperty(HomeKitProperty):
    """Property type for HomeKit smart bulbs."""

    def set_value(self, value):
        """
        Set the current value of the property.

        value -- the value to set
        """
        if self.name != 'color':
            HomeKitProperty.set_value(self, value)
            return

        # Convert color to HSV
        h, s, v = rgb_to_hsv(value)

        # Get the hue, saturation, and brightness properties
        hp = self.device.properties['_hue']
        sp = self.device.properties['_saturation']
        vp = self.device.properties['level']

        # Set all 3 values at once
        ret = self.device.client.set_characteristics({
            '{}.{}'.format(hp.aid, hp.iid): h,
            '{}.{}'.format(sp.aid, sp.iid): s,
            '{}.{}'.format(vp.aid, vp.iid): v,
        })

        if ret:
            self.set_cached_value(value)
            self.device.notify_property_changed(self)

    def update(self, characteristics):
        """
        Update the current value, if necessary.

        characteristics -- current characteristic list for the device
        """
        if self.name.startswith('_'):
            return

        if self.name != 'color':
            HomeKitProperty.update(self, characteristics)
            return

        # Get the hue property and characteristic
        hp = self.device.properties['_hue']
        hc = list(filter(lambda x: x['aid'] == hp.aid and x['iid'] == hp.iid,
                         characteristics))

        # Get the saturation property and characteristic
        sp = self.device.properties['_saturation']
        sc = list(filter(lambda x: x['aid'] == sp.aid and x['iid'] == sp.iid,
                         characteristics))

        # Get the brightness property and characteristic
        vp = self.device.properties['level']
        vc = list(filter(lambda x: x['aid'] == vp.aid and x['iid'] == vp.iid,
                         characteristics))

        if not hc or not sc or not vc:
            return

        value = hsv_to_rgb(hc[0]['value'], sc[0]['value'], vc[0]['value'])
        if value != self.value:
            self.set_cached_value(value)
            self.device.notify_property_changed(self)

File: lib/homekit-property.js
/**
 * HomeKit property type.
 */
'use strict';

const {GattUtils} = require('hap-controller');
const {Property} = require('gateway-addon');

/**
 * HomeKit property type.
 */
class HomeKitProperty extends Property {
  /**
   * Initialize the object.
   *
   * @param {Object} device - Device this property belongs to
   * @param {string} name - Name of the property
   * @param {number} aid - Accessory ID
   * @param {string} serviceType - Service UUID
   * @param {string} characteristicType - Characteristic UUID
   * @param {number} iid - Instance ID
   * @param {string} hapFormat - HAP format
   * @param {Object} description - Description of the property
   * @param {*} value - Current value of this property
   * @param {function?} fromHap - Function to call to transform received value
   * @param {function?} toHap - Function to call to transform new value
   */
  constructor(device, name, aid, serviceType, characteristicType, iid,
              hapFormat, description, value, fromHap = null, toHap = null) {
    super(device, name, description);
    this.aid = aid;
    this.iid = iid;
    this.serviceType = serviceType;
    this.characteristicType = characteristicType;
    this.hapFormat = hapFormat;
    this.fromHap = fromHap;
    this.toHap = toHap;
    this.metaProperty = false;
    this.postUpdate = false;

    this.setCachedValue(value);
  }

  /**
   * Set the current property value, making adjustments as necessary.
   *
   * @param {*} value - New value
   */
  setCachedValue(value) {
    if (this.fromHap && typeof value !== 'undefined') {
      value = this.fromHap(value);
    }

    this.value = value;

    if (this.postUpdate) {
      this.postUpdate(value);
    }
  }

  /**
   * Set the new value of the property.
   *
   * @param {*} value - New value
   * @returns {Promise} Promise which resolves when the value has been set.
   */
  setValue(value) {
    if (this.readOnly) {
      return Promise.reject('Read-only property');
    }

    if (this.hasOwnProperty('minimum')) {
      value = Math.max(this.minimum, value);
    }

    if (this.hasOwnProperty('maximum')) {
      value = Math.min(this.maximum, value);
    }

    if (this.hasOwnProperty('multipleOf')) {
      value = Math.round(value / this.multipleOf) * this.multipleOf;
    }

    if (this.type === 'integer') {
      value = Math.round(value);
    }

    if (this.toHap) {
      value = this.toHap(value);
    }

    if (!this.metaProperty) {
      let client, adapter;

      if (this.device.bridge === null) {
        client = this.device.client;
        adapter = this.device.adapter;
      } else {
        client = this.device.bridge.client;
        adapter = this.device.bridge.adapter;
      }

      switch (this.device.connectionType) {
        case 'ip':
          return client.setCharacteristics({
            [`${this.aid}.${this.iid}`]: value,
          }).then(() => {
            this.setCachedValue(value);
            this.device.notifyPropertyChanged(this);
          });
        case 'ble':
          return adapter.queueBLEOperation(() => {
            return client.setCharacteristics([
              {
                serviceUuid: this.serviceType,
                characteristicUuid: this.characteristicType,
                iid: this.iid,
                value: GattUtils.valueToBuffer(value, this.hapFormat),
              },
            ]).then(() => {
              this.setCachedValue(value);
              this.device.notifyPropertyChanged(this);
            });
          });
        default:
          throw new Error('Unknown connection type');
      }
    } else {
      this.setCachedValue(value);
      this.device.notifyPropertyChanged(this);
      return Promise.resolve();
    }
  }
}

module.exports = HomeKitProperty;
